### PR TITLE
Milestone 3: Game logic

### PR DESCRIPTION
**Pull request by SarvarKh**
### In this milestone we:

**1**. Created a separate branch with the dedicated ruby file for game logic;
**2**. Added the following sections:
> _User input validity check; winner check; end game with restart option;_

**3**. Modified user interaction part

### CODE REVIEW

**Comment by SarvarKh** [comment]
### :heavy_exclamation_mark: For the attention of TSE
I am submitting this PR for review because I exceed a time estimation given to this milestone (:clock2: 14 hours).
But, the game runs as it should :warning:.
_There is definitely a better way to code this game logic._ :dart:

> :bangbang:  _Linter is giving an error for attr_readable. If I rectify it I will have to refractor many things (starting from creating objects, continuing replacing code with them - instead of calling Class directly)_

File: lib/game_logic.rb [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    @board[selected_cell - 1] = user_turn
+
+    puts ''
+    print @board[0..2]
+    puts ''
+    print @board[3..5]
+    puts ''
+    print @board[6..8]
+  end
+end
```
Comment (peterrobert) [inline lib/game_logic.rb]: There should be no puts or print outside the main.rb file.
only_main_print

File: lib/game_logic.rb [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    board[move - 1] = move
+  end
+
+  def self.check_win_combo(board, user)
+    i = 0
+    winner_print = "\n\n\n**************\n**          **\n** #{user} won! **\n**          **\n**************\n\n"
+    while i < @winning_combinations.size
+      if @winning_combinations[i].all? { |number| board.include? number }
+        puts winner_print
+        exit
```
Comment (peterrobert) [inline lib/game_logic.rb]: There should be no puts or print outside of main.rb file.
only_main_print

File: bin/main [hunk 3]
```diff
@@ -1,134 +1,139 @@
 #!/usr/bin/env ruby
 
+require './lib/game_logic.rb'
+
```
Comment (peterrobert) [inline bin/main]: ### [OPTIONAL] 
Using require_relative is in preference to using require. Kindly use require_relative to link your files

**Review by peterrobert — CHANGES_REQUESTED** [review]
### CHANGES REQUIRED ♻️  🛑 
HELLO, @SarvarKh  VERY NICE PROJECT AND GREAT WORK WORKING HARD TO SUBMIT THE PROJECT AT THE REQUIRED TIME. 👏🏿  JUST SOME FEW CHANGES AND THE PROJECT WILL BE GOOD TO GO. LET US HAVE A LOOK ⬇️ 

- [ ] The application does not allow a player to move into an occupied position. At the moment, a player can occupy a position that has already been taken. Kindly this should not be the case.

- [ ] At the moment your game detects a  draw too early, In fact, there are some unoccupied positions. This should not be the case because there is a chance for a player to win. Kindly fix this.

- [ ] When you restart the game, the table is not clearing. Kindly add a method to clear the table if a restart is chosen.

- [ ] In your Lib directory you can remove the `.gitKeep` file. It only needed if there is no file.

- [ ] There are some rubocop offenses. Kindly fix all the rubocop offenses ⬇️ 
Screenshot from 2020-12-11 13-31-38

- [ ] The linter is failing. Kindly fix all the linter errors.

### if you have any question, kindly reach me on slack @peterrobert

### HAPPY CODING 💻

**Review by mahmoud717 — CHANGES_REQUESTED** [review]
### Changes required ♻️ 

**Hello @SarvarKh**

Great job with the project so far it looks amazing, but there are still some changes to be made:
 
---
### Requested 🔴 

- [ ] on every turn, you should show the board with numbers as a reference.


---

please make these changes and request another code review.

if you have any questions don't hesitate to contact me on slack **MahmoudMohammad**

**Review by jubaan — CHANGES_REQUESTED** [review]
# Status: Changes Required :recycle: 

There are some issues that need to be addressed before we can move forward. Please review the list below.

## Required changes :wrench: 
### [Mandatory] :red_circle: 

- [ ] The is stuck at "If you want to see the instructions press any key". Pressing any is not showing anything and after pressing enter the game exits with this error message
image

- [ ] The game is trying to use a `reference_board` that is nonexistent. 

- [ ] The project as it is right now, is not being able to do any of the next features:
  - [ ] `[REQUIRED]` game tells which player turn it is (without computing who the winner is)
  - [ ] `[REQUIRED]` game asks the player to select from available moves
  - [ ] `[REQUIRED]` game informs player if the selected move is invalid (without computing what an invalid move is)
  - [ ] `[REQUIRED]` game displays board after the player move (without computing which spots on the board are taken)
  - [ ] `[REQUIRED]` game informs player if the selected move is a winning move (without computing the winning move)
  - [ ] `[REQUIRED]` game informs player if the selected move is a draw move (without computing draw)
  - [ ] `[REQUIRED]` game repeats all actions for the next player's move

## Keep up the great work :muscle: and Happy Coding :smiley: :computer:!!

**Comment by SarvarKh** [comment]
### Hey, @jubaan thank you for your review. 
**I simply didn't push the last version of my code which contained reference_board** :sweat_smile: and that's the reason why the game is not running))
_Sorry about that._

**Comment by descholar-ceo** [comment]
# This Pull Request is under review by [descholar-ceo](https://github.com/descholar-ceo) :eyes: :eyes: 
## Please wait a bit ...

**Review by descholar-ceo — CHANGES_REQUESTED** [review]
# Status: Changes Required :vertical_traffic_light: :warning: 
## Hello @SarvarKh  :handshake: :handshake: :handshake: 

:+1: This is good to make it this far, your project is really nice, congratulations on the good work you have done so far.
#### :warning: However we need to make few changes to mark this project approved, don't feel discouraged because of being asked some changes more and more, that's beacuse this is a learning process, and we are here to help you to become an out-standing developer. See the requested change below :point_down: :point_down: :point_down: 
### Note: All of requested changes which are marked as [Required] are mandatory to be implemented in order to get this project approved, others which are marked as [Optional] are nice to have in your project to make it outstanding

# [Required Changes] :red_circle: 
- [ ] When I enter an invalid move `alphabet` or `a special character`, at first it accepts and check the place 9, that is a bug please fix it.

| When I press alphabet | When I press a special character |
| --- | --- |
|Screenshot from 2020-12-14 23-22-05|Screenshot from 2020-12-14 23-22-42|

- [ ] When there is a draw or a win, the game should ask players if they want to play another round, it gives them two options, option 1: if they enter Y or yes, the game will continue without termination, option 2: the game will be terminated if they press n or no, or any other key

# [Optional changes] :yellow_circle: 
:warning: Your commit messages should be in imperative mode or present tense, _i.e.:_ you should have written `Add method to check winning combs` instead of writing `Added method to check winning combs`, refer [to this article for more details on how to write a good && professional commit message](https://www.freecodecamp.org/news/level-you-up-to-awesome-commit-messages-a85558cb90e8)

___
_If anything seems to be unclear to you, reachout and ask anything by commenting on this Pull Request and mention me as [descholar-ceo](https://github.com/descholar-ceo), I will receive a notification and will respond._

[GitHub](https://github.com/descholar-ceo) | [LinkedIn](https://linkedin.com/in/mugirase-emmanuel) | [Twitter](https://twitter.com/descholar3)

_Reviewed by [Mugirase Emmanuel (descholar)](https://github.com/descholar-ceo)_

**Comment by descholar-ceo** [comment]
# This Pull Request is under review by [descholar-ceo](https://github.com/descholar-ceo) :eyes: :eyes: 
## Please wait a bit ...

File: bin/main.rb [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    my_board.move(move, current_user.sign, my_board)
+    puts my_board.display
+
+    found_winner = my_board.check_winner(my_board.boardx, my_board.boardo, current_user.name)
+    if found_winner
+      puts "#{found_winner} won!"
+      puts 'Do you want to play another round?'
+      puts "Enter 'Y' or 'yes' if you want to continue. Otherwise press any key to end the game!"
+      end_game = gets.chomp
+      %w[yes Y].include?(end_game) ? my_board.turn_counter = 8 : exit
```
Comment (descholar-ceo) [inline bin/main.rb]: I would suggest that here you add `y` so if the user press `Y` will be the same as pressing `y`, remember that some users don't care whether you said an uppercase or lowercase letter in instructions

**Review by descholar-ceo — CHANGES_REQUESTED** [review]
# Status: Changes Required :vertical_traffic_light: :warning: 
## Hello @SarvarKh  :handshake: :handshake: :handshake: 

:+1: This is good to make it this far, your project is really nice, congratulations on the good work you have done so far.
:+1: I liked that you have tried to implemented the requested changes from the previous review, keep it up, and your project is really almost done.
:+1: Great on implementing the `other round?` feature I suggested in the previous review.
#### :warning: However we need to make few changes to mark this project approved, don't feel discouraged because of being asked some changes more and more, that's beacuse this is a learning process, and we are here to help you to become an out-standing developer. See the requested change below :point_down: :point_down: :point_down: 
### Note: All of requested changes which are marked as [Required] are mandatory to be implemented in order to get this project approved, others which are marked as [Optional] are nice to have in your project to make it outstanding

# [Required Changes] :red_circle: 
>  - [ ] When I enter an invalid move `alphabet` or `a special character`, at first it accepts and check the place 9, that is a bug please fix it.

| When I press alphabet | When I press a special character |
| --- | --- |
|Screenshot from 2020-12-14 23-22-05|Screenshot from 2020-12-14 23-22-42|

- [ ] The above :point_up: :point_up: :point_up: issue is still there. Please make sure that you fix it before you submit for a re-review. 
_Note:_ If you don't get it well how it appears, please contact me as I stated it below :point_down: 
___
_If anything seems to be unclear to you, reachout and ask anything by commenting on this Pull Request and mention me as [descholar-ceo](https://github.com/descholar-ceo), I will receive a notification and will respond._

[GitHub](https://github.com/descholar-ceo) | [LinkedIn](https://linkedin.com/in/mugirase-emmanuel) | [Twitter](https://twitter.com/descholar3)

_Reviewed by [Mugirase Emmanuel (descholar)](https://github.com/descholar-ceo)_

**Review by Srjnnnn — APPROVED** [review]
## Status Approved :partying_face:  :tada: 
Hello @SarvarKh. Thanks for making the requested changes. Your game has passed all the behavioral and implementation features that it needs to have. :muscle: 

There is one **OPTIONAL** change I'd ask from you.

- [ ] In every turn your game puts the number table. I think showing the latest state is enough and will be better for the user experience. 
Since this is gonna be your showcase, I highly recommend you implement this requested change. Other than that, you can merge this PR and move on to the next one. :+1:  
_Eyüp Sercan UYGUR_

**Comment by SarvarKh** [comment]
### Hello @eypsrcnuygr. Thank you for your swift feedback and recommendation 👍 👍 
I do agree with you on that. Originally there was no such feature. But, earlier TSE had requested me to add. As rule, I did what was requested.
I will remove this method from each turn. Showing it at the beginning shall be enough.

**Comment by Srjnnnn** [comment]
I believe he or she meant that you need to show the game board with the latest state. However, there may be confusion also. So show the game state on every turn but there is no need for all numbers. Good luck and happy coding.

**Comment by SarvarKh** [comment]
Yes, maybe...

Anyway, thank you for your review.

**Pull request merged**